komga/src/main/kotlin/org/gotson/komga/interfaces/rest/SeriesController.kt

```diff
@@ -319,17 +325,77 @@ class SeriesController(
     } ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
 
   @ApiResponse(content = [Content(schema = Schema(type = "string", format = "binary"))])
-  @GetMapping(value = ["{seriesId}/thumbnail"], produces = [MediaType.IMAGE_JPEG_VALUE])
+  @GetMapping(value = ["{seriesId}/thumbnail", "{seriesId}/thumbnails/{thumbnailId}"], produces = [MediaType.IMAGE_JPEG_VALUE])
   fun getSeriesThumbnail(
     @AuthenticationPrincipal principal: KomgaPrincipal,
-    @PathVariable(name = "seriesId") seriesId: String
+    @PathVariable(name = "seriesId") seriesId: String,
+    @PathVariable(name = "thumbnailId") thumbnailId: String?
   ): ByteArray {
-    seriesRepository.getLibraryId(seriesId)?.let {
-      if (!principal.user.canAccessLibrary(it)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
-    } ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+    val series = seriesRepository.getLibraryId(seriesId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+    if (!principal.user.canAccessLibrary(series)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
+
+    val thumbnail = if (thumbnailId != null) {
+      seriesLifecycle.getThumbnailBytesById(thumbnailId)
+    } else {
+      seriesLifecycle.getThumbnailBytes(seriesId, principal.user.id)
+    }
+    return thumbnail ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+  }
+
+  @GetMapping(value = ["{seriesId}/thumbnails"], produces = [MediaType.APPLICATION_JSON_VALUE])
+  fun getSeriesThumbnails(
+    @AuthenticationPrincipal principal: KomgaPrincipal,
+    @PathVariable(name = "seriesId") seriesId: String
+  ): Collection<SeriesThumbnailDto> {
+    seriesRepository.findByIdOrNull(seriesId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+    if (!principal.user.canAccessLibrary(seriesId)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
+
+    return thumbnailSeriesRepository.findAllBySeriesId(seriesId)
+      .map { it.toDto() }
+  }
+
+  @PostMapping(value = ["{seriesId}/thumbnail"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
+  @PreAuthorize("hasRole('$ROLE_ADMIN')")
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  fun postUserUploadedSeriesThumbnail(
+    @PathVariable(name = "seriesId") seriesId: String,
+    @RequestParam("image") image: MultipartFile
+  ) {
+    val series = seriesRepository.findByIdOrNull(seriesId) ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+
+    val inputStream = image.inputStream
+    val bufferedInputStream = inputStream.buffered()
+
+    val type = contentDetector.detectMediaType(bufferedInputStream)
+    if (!contentDetector.isImage(type)) {
+      throw ResponseStatusException(HttpStatus.UNSUPPORTED_MEDIA_TYPE)
+    }
+    seriesLifecycle.addUserUploadedThumbnailForSeries(series, image.bytes)
 
-    return seriesLifecycle.getThumbnailBytes(seriesId, principal.user.id)
-      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)
+    inputStream.close()
+    bufferedInputStream.close()
```

Comment:
You close the stream quite late, an exception could have been thrown before. The `use {}` syntax is preferred for consistency